### ISSUE TITLE
Port twisted.internet.fdesc to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5974]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5974|
|Type|enhancement|
|Created|2012-09-13 18:26:17Z|
|Branch|https://github.com/twisted/twisted/tree/fdesc-py3-5974|

`twisted.internet.fdesc` should run on Python 3.

Attachments:

* [fdesc.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4bd/4bd80e814010cc488c1e8a508b75ab19d1f4083c/e98d40c79c74f721f11ac3e64be3a95387b04182.diff) (6634 bytes) - added by itamar on 2012-09-13 18:40:23Z - Port fdesc; depends on #5973

<details><summary>Searchable metadata</summary>

```
trac-id__5974 5974
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_fdesc_py3_5974 branches/fdesc-py3-5974
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347560777000000 1347560777000000
changetime__1347657786000000 1347657786000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

This depends on [#5973](https://github.com/twisted/twisted/issues/5973); I will implement it locally and branch when that is merged.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Blocking is a property of file descriptions, not file descriptors.  At least the test method docstrings get this wrong.
  1. Existing defect, but `ReadWriteTestCase` docstring could use `L{}` and the `setUp` docstring says it creates two pipes when it only creates one.
  1. `test_setCloseOnExec` and `test_unsetCloseOnExec` leak the file they open.

Guess I'll look again after [#5973](https://github.com/twisted/twisted/issues/5973) is resolved.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [35666]) Branching to 'fdesc-py3-5974'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

I've switched docstrings to say file description, fixed docstrings, and made sure files don't leak. Started build run:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fdesc-py3-5974

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set status to `closed`|
|-|-|

(In [35683]) Merge fdesc-py3-5974.

Author: itamar
Review: exarkun
Fixes: [#5974](https://github.com/twisted/twisted/issues/5974)

Port twisted.internet.fdesc to Python 3.
